### ISSUE TITLE
Not working with Bluetooth Proxies (ESP32 based)

### ISSUE THREAD

**Issue by bcutter** [GrumpyMeow/sonicare-ble-hacs#14]
While I was lucky to initialy set up this integration without any issues (https://github.com/GrumpyMeow/sonicare-ble-hacs/issues/2#issuecomment-1715326758), I discovered by manually reloading the integration that:
- it will find the toothbrush just fine if it is pretty close to the Home Assistant server - a Raspberry Pi 4 which uses native Bluetooth
- it will NEVER EVER find the toothbrush if next to a Bluetooth Proxy (ESP powered, https://esphome.github.io/bluetooth-proxies/) - at the same time it can be discovered using the "LightBlue" app on an iPhone also next to the BT Proxy ESP

**So where's the issue?** Is it the BT proxy or this Sonicare BLE integration?

I bet this is one of the reasons for the most open "Can't connect to XYZ" issues in this integration, so really worth looking into this @GrumpyMeow! We should probably start asking for the specific bluetooth environment people use. I already found few users stating they are also using Bluetooth Proxy (e. g. here: https://github.com/GrumpyMeow/sonicare-ble-hacs/issues/2#issuecomment-1556674265).

**Comment by badewanne1234** [comment]
Thanks for sharing this. As stated, I can see the advertising messages of the Toothbrush in the BT Proxy Log, so I think it might be an issue of how the Sonicare integration reads/interprets these advertisement messages OR is the raspi BT passing BT messages differently than the ESP32 based BT proxies?

**Comment by bcutter** [comment]
> it might be an issue of how the Sonicare integration reads/interprets these advertisement messages OR is the raspi BT passing BT messages differently than the ESP32 based BT proxies?

That's exactly the question I hope bluetooth experts like @GrumpyMeow can verify.

By the way: by only activating my toothbrush next to the BT proxy I could not find any activity in the logs (ESPHome). If i trigger other bluetooth devices (like SwitchBot buttons) next to the BT proxy, I see quite some activity.

**Comment by badewanne1234** [comment]
I think I played around with the advertising window / intervalls and could see the activity immediately when lifting the TB out of the glass...

**Comment by edwardtfn** [comment]
So, I've teste getting quite close to the server (like 2m) and the toothbrush was finally discovered. 😃 

Now I will vote to add support for Bluetooth proxy. 😉

**Comment by bcutter** [comment]
@GrumpyMeow

**Comment by PV-Web** [comment]
Same problem. Bluetooth proxy via the M5 atom lite via ESP home. @GrumpyMeow 
Hopefully you got time to check were this could go wrong.

**Comment by helmerzNL** [comment]
My toothbrush does connect if i'm close to my HA instance, but not from a Bluetooth Proxy. So, it would be great if support for Bluetooth proxies can be added.

**Comment by bcutter** [comment]
@GrumpyMeow are you reading/still maintaining this project?

**Comment by rigwig** [comment]
Same issue here, am able to connect via HA intgrated bt, but no luck via an esp32 based proxy.

**Comment by bcutter** [comment]
> @GrumpyMeow are you reading/still maintaining this project?

@GrumpyMeow you seem to be active, unfortunately not in this repo. Can you please give us an update on this? Fixing this "does not work with bluetooth proxies" issue is a mandatory killer feature immediately resolving a ton of open issues.

**Comment by urosch** [comment]
This is the log output from home assistant after trying to connect my sonicare to ha via esphome:
`2024-01-07 12:00:28.419 INFO (SyncWorker_7) [homeassistant.util.package] Attempting install of sonicare-bletb==0.0.8
2024-01-07 12:01:28.647 WARNING (MainThread) [sonicare_bletb.sonicare_bletb] Philips Sonicare: Init
2024-01-07 12:01:28.647 WARNING (MainThread) [sonicare_bletb.sonicare_bletb] Philips Sonicare: initialise
2024-01-07 12:01:28.647 WARNING (MainThread) [sonicare_bletb.sonicare_bletb] Philips Sonicare: Connecting; RSSI: None
2024-01-07 12:02:11.702 WARNING (MainThread) [sonicare_bletb.sonicare_bletb] Philips Sonicare: Init
2024-01-07 12:02:11.702 WARNING (MainThread) [sonicare_bletb.sonicare_bletb] Philips Sonicare: initialise
2024-01-07 12:02:11.702 WARNING (MainThread) [sonicare_bletb.sonicare_bletb] Philips Sonicare: Connecting; RSSI: None
2024-01-07 12:03:29.214 WARNING (MainThread) [sonicare_bletb.sonicare_bletb] Philips Sonicare: Init
2024-01-07 12:03:29.214 WARNING (MainThread) [sonicare_bletb.sonicare_bletb] Philips Sonicare: initialise
2024-01-07 12:03:29.214 WARNING (MainThread) [sonicare_bletb.sonicare_bletb] Philips Sonicare: Connecting; RSSI: None
2024-01-07 12:03:49.238 WARNING (MainThread) [aioesphomeapi.connection] esphome-web- @ 100.0.0.1: Connection error occurred: [Errno 104] Connection reset by peer
2024-01-07 12:03:49.238 INFO (MainThread) [aioesphomeapi.reconnect_logic] Processing unexpected disconnect from ESPHome API for esphome-web- @ 100.0.0.1
2024-01-07 12:03:49.251 INFO (MainThread) [aioesphomeapi.reconnect_logic] Successfully connected to esphome-web- @ 100.0.0.1 in 0.005s
2024-01-07 12:03:49.329 INFO (MainThread) [aioesphomeapi.reconnect_logic] Successful handshake with esphome-web- @ 100.0.0.1 in 0.078s`

**Comment by adheaps** [comment]
Seems @GrumpyMeow has abandoned this project. Would love to see some progress on this. Is this something that the Home Assistant dev team would pick up? It appears to have some good interest from the community.

**Comment by bcutter** [comment]
I don't think that is gonna happen. Usually a custom integration needs to prove it is "good enough" (like function in general, reliable, flexible in terms of reaction time for changes in core etc.) to become a core/default integration.

Very sad here's no progress, which makes this basically really great integration basically completely useless for most users. Sell Philips, get a Oral-B? Just kidding.

**Comment by jlipsit** [comment]
### Summary of Errors and Troubleshooting Steps for ESP32 Bluetooth Proxy with Sonicare BLE Toothbrush

_Updated: I found that although it will connect to a BT Dongle directly connected to the controller, it does not reconnect reliably after the toothbrush goes to sleep. I have to RELOAD the integration when the brush is awake to get it to reconnect. It only stays connected to that "session" and then will not reconnect. Therefore I suspect the integration has some issues that are just more severe when used with a proxy. **IS ANYONE USING THIS RELIABLY WITH A DIRECT CONNECTION TO A BT DONGLE?**_

I picked up a new Philips Sonicare and verified that it will not provision and connect when using a BT proxy, It will connect and operate properly when using a BT dongle connect to the Home Assistant Green controller. 

Although the toothbrush is discovered successfully, when attempting to add it to HA, when using an ESP32 Bluetooth proxy  the Sonicare BLE toothbrush disconnects with error messages indicating "Connection reset by peer" (Errno 104). This issue occurs consistently, preventing the ability to add it to HA.

I did a lot of troubleshooting............... Here are some of my logs and troubleshooting

**ESP32 Device Log Extract:**

BLE Connection Attempts and Errors:

**First Connection Attempt:**
Timestamp: [08:05:24]
Device MAC: D3:35:34:35:53:8C
Event:
[08:05:24] Attempting BLE connection
[08:05:27] ESP_GATTC_CONNECT_EVT
[08:05:27] ESP_GATTC_OPEN_EVT
[08:05:27] Connected
[08:05:36] Disconnecting
[08:05:36] ESP_GATTC_CLOSE_EVT

**Second Connection Attempt:**
Timestamp: [08:06:42]
Device MAC: 24:E5:AA:75:27
Event:
[08:06:42] Attempting BLE connection
[08:06:42] ESP_GATTC_CONNECT_EVT
[08:06:42] ESP_GATTC_OPEN_EVT
[08:06:56] ESP_GATTC_DISCONNECT_EVT, reason 8
[08:06:56] ESP_GATTC_CLOSE_EVT
[08:06:57] Attempting BLE connection
[08:06:58] ESP_GATTC_CONNECT_EVT
[08:06:58] ESP_GATTC_OPEN_EVT

**Third Connection Attempt:**
Timestamp: [08:07:26]
Device MAC: 24:E5:AA:75:27
Event:
[08:07:26] Attempting BLE connection
[08:07:33] ESP_GATTC_CONNECT_EVT
[08:07:33] ESP_GATTC_OPEN_EVT
[08:07:46] Disconnecting
[08:07:47] ESP_GATTC_CLOSE_EVT
[08:07:47] Attempting BLE connection
[08:07:48] ESP_GATTC_CONNECT_EVT
[08:07:48] ESP_GATTC_OPEN_EVT

**Errors:**
Timestamp: [08:06:42]
Error: hcif disc complete: hdl 0x0, rsn 0x3e
Timestamp: [08:06:56]
Error: gattc_conn_cb: if=3 st=0 id=3 rsn=0x8
Timestamp: [08:06:57]
Error: BTM_BleScan scan not active
Error: bta_dm_ble_scan stop scan failed, status=0x6
Timestamp: [08:07:26]
Error: hcif disc complete: hdl 0x0, rsn 0x3e
Timestamp: [08:07:46]
Error: GATTC_ConfigureMTU GATT_BUSY conn_id = 3
Error: hcif disc complete: hdl 0x0, rsn:0x13
Error: No pending command

**Summary:**
The ESP32 attempts to connect to the BLE device multiple times, experiencing intermittent success and errors such as disconnection events and scan failures. Each connection attempt and its outcome are logged with timestamps for detailed troubleshooting.

**Successful Initial Connections:**

- The log shows successful connection attempts to the toothbrush's BLE device, identified by its MAC address (e.g.,  [24:E5:AA:75:27:D9]).
- BLE events such as ESP_GATTC_CONNECT_EVT and ESP_GATTC_OPEN_EVT indicate that the ESP32 Bluetooth proxy is able to establish initial connections.
- the signal strength appears adequate (Signal strength: -58 dB)

**Frequent Disconnections:**

- The log contains several instances of disconnection events (ESP_GATTC_DISCONNECT_EVT), often followed by immediate reconnection attempts.
- Disconnection reasons (rsn) vary but commonly include 0x16 and 0x3e.

**Warnings and Errors:**

- Warnings such as Connection error occurred: [Errno 104] Connection reset by peer suggest that the connection is being unexpectedly terminated.
-  There are several instances of hcif disc complete: hdl 0x0, rsn 0x16 indicating that the disconnection is being completed by the Bluetooth host.





**Troubleshooting Steps Taken:**

**Configuration Adjustments:**

**Adjusted esp32_ble_tracker scan parameters to optimize BLE scanning:**

```
esp32_ble_tracker:
  scan_parameters:
    interval: 320ms
    window: 160ms
```

```
esp32_ble_tracker:
  scan_parameters:
    interval: 160ms
    window: 80ms
```




**Power Supply Verification:**

Ensured that the ESP32 is connected to a stable power source to avoid power-related disconnections.
Interference Minimization:

Reduced potential Bluetooth and Wi-Fi interference by positioning the ESP32 and Sonicare toothbrush closer together and away from other electronic devices.

**Firmware and Software Updates:**

Confirmed that all firmware and software (ESPHome, Home Assistant, and ESP32 firmware) are updated to the latest versions available at the time of testing.

Core
2024.7.1
Supervisor
2024.06.2
Operating System
12.4
Frontend
20240705.0


**Logger Configuration:**

**Enabled detailed logging to capture more granular details for debugging:**
```
logger:
  level: debug
  logs:
    esp32_ble_client: debug
```




### Alternative Connectivity Test:

**Verified that the Sonicare toothbrush connects successfully using a Bluetooth dongle plugged directly into the Home Assistant Green controller, indicating the issue is specific to the ESP32 Bluetooth proxy.**

![image](https://github.com/GrumpyMeow/sonicare-ble-hacs/assets/155367039/f1454b34-56bb-4600-a550-06810d56ffaf)

**Comment by bcutter** [comment]
Great work! For sure took a loooooot of time. How should we use those perfect information (you mentioned "the ESP home guys need to look at this issue")?

Maybe someone can ping them?

Another note: I also use Shelly's as bluetooth proxies, with the same (not working) results as with a ESP home powered bluetooth proxy. Shelly are also based on ESP, but not ESP Home - just as a small note.

**Comment by jlipsit** [comment]
Unfortunately, I don't think they will look at it until someone takes over this integration and reviews the code for possible errors.  It's sad because the Oral B is very expensive and the Sonicare has excellent hardware and is very affordable.

**Comment by jlipsit** [comment]
I found that although it will connect to a BT Dongle directly connected to the controller, it does not reconnect reliably after the toothbrush goes to sleep. I have to RELOAD the integration when the brush is awake to get it to reconnect. It only stays connected to that "session" and then will not reconnect. Therefore I suspect the integration has some issues that are just more severe when used with a proxy. **IS ANYONE USING THIS RELIABLY WITH A DIRECT CONNECTION TO A BT DONGLE?**_

**Comment by edwardtfn** [comment]
> **IS ANYONE USING THIS RELIABLY WITH A DIRECT CONNECTION TO A BT DONGLE?**

Unfortunately, no. 😩

**Comment by jlipsit** [comment]
> > **IS ANYONE USING THIS RELIABLY WITH A DIRECT CONNECTION TO A BT DONGLE?**
> 
> Unfortunately, no. 😩

So just to clarify, you have it connected to a BT dongle directly, but it is not reliable?

**Comment by edwardtfn** [comment]
Exactly. In the end, it's not useful.

**Comment by Dodoooh** [comment]
is there any good news to get this up and running?

**Comment by bcutter** [comment]
No. The author of this integration doesn't seem to be active in any way. This makes us (potential) users grumpy and leaves us with a non-working integration. Completely useless.

**Comment by v6ak** [comment]
I remember having similar issues when implementing QuietDrift for SwitchBot Curtain 3. Originally, it was a separate service called via HTTP from HA. However, I got into connection issues similar to those. I implemented various reconnecting mechanisms, but it was hard to make it working in long term. Then, I rewrote it as a custom component and let HA to manage the BT connection, which has fixed the connection issues.

My original code was obtaining the device object from bleak_retry_connector.get_device. And this this custom component does something similar:

https://github.com/GrumpyMeow/sonicare-ble-hacs/blob/5ab0c8fa3c0a44015a8b75fa7196a89ada48b95e/custom_components/sonicare_bletb/__init__.py#L27-L29 

Well… It isn't exactly the same. It tries to use [bluetooth.async_ble_device_from_address](https://developers.home-assistant.io/docs/core/bluetooth/api/) first.  However, when the device isn't reachable, it returns None, which forces a fallback to bleak_retry_connector.get_device. And it is quite common that the device is unreachable. In this case, bleak_retry_connector.get_device is likely to also fail, but it doesn't have to, as it is called somewhat later, and apparently can wait a bit. In my logs, I've seen an error from bleak_retry_connector, which suggests that this sometimes happens.

The code is in async_setup_entry, and it is the only place where async_ble_device_from_address/get_device is called, which has few consequences:

* It usually fails on startup, unless the toothbrush is reachable. (It is rarely the case.)
* It seems HA retries later when it discovers such device.
* Once the connection is lost, there is probably no mechanism to recover from that. HA will probably still consider the setup entry as valid, but the component doesn't seem to have a mechanism for reconnecting.

Not sure about async_ble_device_from_address, but I am pretty sure that bleak_retry_connector.get_device retries just in the first connection. Once the connection is lost, the returned object doesn't retry.

Initially, I am removing the get_device fallback and will see how it affects the behavior. I am not sure why it was added, as 34508e38cf81532f29b787d390f814259c9397f5 doesn't explain much.

**Comment by v6ak** [comment]
Few more findings:

First, it has connected to the toothbrush even with all BT adapters disabled, leaving only BT proxies.

Second, it has still issues with disconnecting. Looking more about the code of this repo, [the library](https://github.com/GrumpyMeow/sonicare-ble) and the documentation, it seems that the component (and/or its libraries) is responsible for reconnections. There is however some mechanism, but seemingly incomplete:

1. The library has [_reconnect method](https://github.com/GrumpyMeow/sonicare-ble/blob/8a7969b046ed153e687b15bfa468940d69db7f0b/src/sonicare_bletb/sonicare_bletb.py#L326-L337), which is called from [_disconnected](https://github.com/GrumpyMeow/sonicare-ble/blob/8a7969b046ed153e687b15bfa468940d69db7f0b/src/sonicare_bletb/sonicare_bletb.py#L345-L358) and from itself. However, calling from itself happens only in case of `except BleakNotFoundError:`,  not in case of other errors. But the errors are more diverse.
2. The component [listens for new device with the toothbrush's address](https://github.com/GrumpyMeow/sonicare-ble-hacs/blob/5ab0c8fa3c0a44015a8b75fa7196a89ada48b95e/custom_components/sonicare_bletb/__init__.py#L45-L63) and calls [set_ble_device_and_advertisement_data](https://github.com/GrumpyMeow/sonicare-ble/blob/8a7969b046ed153e687b15bfa468940d69db7f0b/src/sonicare_bletb/sonicare_bletb.py#L85-L96), which, aside of logging, just sets properties, but doesn't call _reconnect. Which means it cannot save us when the reconnect loop is broken.

Looking from 10 feets:

1. Not sure if bleak_retry_connector.get_device is to blame. It shouldn't be needed, maybe it was added as an attempt to fix something (the megacommit 34508e38cf81532f29b787d390f814259c9397f5 doesn't tell us much). Now, it seems to just add complexity and I'll probably keep it removed.
2. The reconnects depend on polling (quite frequent, 0.25s), but it's a bit weird to poll for a toothbrush, which is offline most of the time.
3. The polling frequency will affect what kind of issues we face. Frequent polling increases likelihood of errors other than BleakNotFoundError (that stop the polling). Less frequent polling increases likelihood that HA will not connect within the short window when toothbrush is online.

What to do:

1. First and foremost, the _async_update_ble should trigger reconnect ASAP. (Not sure what to do when it is still connected.) So we woulnd't rely on polling.
2. We should also register a disconnect handler which stops the polling when toothbrush is disconnected. This would make the polling effectively disabled (without actually removing the polling code from the library).
3. Bonus: don't rely on toothbrush being available on startup (in async_setup_entry).

I'll probably do it later, it doesn't look that hard.

EDIT: I got a bit struggles with the callbacks, HA's documentation isn't perfect. But maybe I was reinventing the wheel – it looks like ActiveBluetoothDataUpdateCoordinator has already implemented what I need.

**Comment by bcutter** [comment]
@v6ak Wow, impressive (as always when users see devs talking and doing their magic 😀).

Sounds like maybe Christmas takes place a bit earlier this year: toothbrush working with BT proxies sounds so 👍

**Comment by simbalasim** [comment]
@v6ak 
Wow, that sounds really promising. I'm not a developer either, but that sounds like you have found at least a direction how to progress further. Much appreciated 
Really great that you're taking this on and investing your time. Many thanks for that 😃,

I'm currently working with very unreliable workournds,. when the toothbrush connects to a tablet, the Soniccare app simply opens. Not particularly fancy ^^

Unfortunately I don't know coding, but if I can help e.g. with testing, please let me know.

**Comment by v6ak** [comment]
Well, not sure how to proceed. I've tried multiple ways to discover the device:

* bluetooth.async_register_callback
* ActiveBluetoothDataUpdateCoordinator
* bluetooth.async_get_scanner

Expected results: the device gets discovered when I make the toothbrush active (move, brushing, briefly after brushing, briefly after charging), so I can connect it.

Actual results: the device gets discovered immediately and seems to never disappear. However, I don't get a callback when the device is actually activated.

I feel like I am missing something trivial, but not sure what…

**Comment by jlipsit** [comment]
With all my testing, I got the impression that there was some bug in the brush itself. I just threw mine in the trash this week and purchased 4 Oral B Series 4 toothbrushes off eBay. Great brush and about $50 bucks new on eBay if you are patient. The are rock solid in Home Assistant. 

Not Running View
![image](https://github.com/user-attachments/assets/34692196-e6b0-4788-8048-48bb32602605)

Running View
![image](https://github.com/user-attachments/assets/8667cf7f-8bec-4ed6-8a0a-fb48a5b65bb3)

**Comment by u20p17** [comment]
@v6ak : thank you very much for trying to solve the issue! Looking forward to test your fixes in the future with my philips toothbrush 👌

Regarding you missing callback: maybe you could as in the discord channel of HomeAssistant in the dev area. Maybe you get some hints 👍

**Comment by adheaps** [comment]
@v6ak - I'd like to thank you as well. So happy to see someone trying to solve this issue. Truly appreciate your efforts on this.

**Comment by v6ak** [comment]
I was trying to perform a deeper rewrite of the BT handling, but ended with just conservative changes:

* Got rid of get_device fallback when HA's API fails. Not 100% sure about its impact, but it should be unneeded.
* When the toothbrush reappears, I force reconnect even if the original reconnect loop (which seems to be unreliable) has died.
* CI, not directly relevant to end users

I don't think it will be 100% reliable, but it might be reliable enough for some uses, including my needs (battery monitoring and keeping the charge about 50 %). My adjustments in reconnect will probably work only after longer inactivity, which might be enough for typical use (brushing twice a day).

Because testing needs longer inactivity, it is not much tested, but if you want to try it anyway, it's there: https://github.com/v6ak/sonicare-ble-hacs

Some issues that aren't fixed:

* When HA boots, the toothbrush will appear as failed until it connects.
* Reconnects might be a bit aggressive, potentially delaying communication with others devices.

## How (re)connecting works, what are the challenges?

It seems to be quite easy to initiate the connection to the toothbrush. However, you probably don't want to press a button (within the short time-frame) in order to do that. It should be done automatically.

The toothbrush is apparently offline for most of the time, so it is crucial to detect when it becomes online. Also, it is crucial to detect is ASAP, because the time when you can connect is quite short.

One could try to connect periodically (so called polling). The polling would have to be quite frequent (as the time for connection is quite short). This would not only take CPU (maybe not that much), it also uses Bluetooth communication. Depending on your BT adapter and other BT devices, you might reach the limit of maximum connected devices. This addon (well, the sonicare-blebt library) does this partially, but it seems to break in long term.

The other way to do it is to somehow detect when a toothbrush is available, just by being passive and waiting for some indication. BT devices announce their presence (if they are configured to be visible), and it seems to be also the case of Sonicare toothbrush when it starts being active (brushing, charging ends, brush moves, …). However, apparently, I can't see all the beacons. [The HA Bluetooth API documentation](https://developers.home-assistant.io/docs/core/bluetooth/api/) suggests that you probably will not receive two such notification in a short succession:

> To get a callback when the Bluetooth stack can no longer see a device, call the bluetooth.async_track_unavailable API. For performance reasons, it may take up to five minutes to get a callback once the device is no longer seen.

Yes, this is about *unavailability* notification, but availability notifications probably work in a similar manner. When HA doesn't see the device, it takes a time to consider it unavailable. When a beacon from the device is received, HA probably doesn't notify about its availability, as it didn't consider the device to be unavailable. In my testing, it looked like there is some longer period than just five minutes, but I might got something wrong or misunderstand something.

## Possible workaround for missing the availability notification

If my changes aren't enough and HA doesn't detect the presence of the toothbrush enough, we could add a service that would connect the toothbrush when called. It would be up to user to call it when needed. This could work automatically in some setups, e.g.:

1. I'll assume that the toothbrush is usually on a charger and the charger is connected through a smart socket, which is off most of the time.
2. On some event, the smart socket would briefly (for 5 seconds or so) turn on and then off, so the toothbrush becomes active.
3. As HA is aware when this happens, some automation could call the service.

I hope this will not be needed, though.

**Comment by u20p17** [comment]
@v6ak : thanks again! today in the morning I briefly tried to connect my sonicare via a bluetooth proxi to HA, but sadly I couldn't connect. I didn't have time to check logs, but in theory the connection via bluetooth proxis should work (with your branch)?

**Comment by v6ak** [comment]
Do you try to add a new toothbrush, or to use an already added toothbrush? Using an already added toothbrush should work. At least, it works for me.

Adding a new toothbrush via BLE proxy? Not sure:

* I haven't tested adding a toothbrush with the modified addon.
* With the original addon, I had to try several (four?) times, even when having BT adapter enabled. Not sure if it went through the adapter or through BLE proxy. (I was about 5m from both proxy and adapter.)
* Also, the opportunity to connect is quite short. When your toothbrush is active (after brushing, after charging, when you move it, …), there is very short period to connect to the toothbrush. I don't think I can do anything about it.

EDIT: I've also found an example of what breaks the original reconnection BleakNotFoundError: bleak_retry_connector.BleakOutOfConnectionSlotsError: […] - AA:DD:55:EE:SS:SS: Failed to connect after 9 attempt(s): No backend with an available connection slot that can reach address AA:DD:55:EE:SS:SS was found: The proxy/adapter is out of connection slots or the device is no longer reachable; Add additional proxies (https://esphome.github.io/bluetooth-proxies/) near this device

**Comment by v6ak** [comment]
It seems that the issue with device discovery was [independently described by GrumpyMeow](https://github.com/GrumpyMeow/sonicare-ble-hacs/issues/11#issuecomment-1961872893). Anyway, it looks like the connection to the toothbrush is successful if and only if the last attempt is old enough (EDIT) or fresh enough, but not few minutes after the last attempt. Not perfect, probably good-enough for me. Not sure if we can do much about it without changing the HA core, except adding a service for forcing a reload.
EDIT: Maybe `bluetooth.async_set_fallback_availability_interval(hass, address, 1.0)` or something similar will do the trick.
EDIT2: Maybe it is not that simple. there seem to be [two different default timeouts](https://github.com/Bluetooth-Devices/habluetooth/blob/7526057b1839b9929a802600fdb5c117465ea1ea/src/habluetooth/const.py#L15-L34). And the timeout for connectable devices doesn't seem to be overridable.
EDIT4: Maybe async_rediscover_address will do the trick.

What's not that good is that only some entities seem to be updated. Handle time seems to be updated, but battery state doesn't. This is kinda weird, as both seem to be implemented in the very same way. I hope I can work around by full reinitialization.

EDIT3: I've also looked on active vs. passive scanning. I think that passive scanning would be faster and enough for the toothbrush connection. However, because [HA actually ignores the mode for now](https://github.com/home-assistant/core/blob/3d26fa7864b612d7686519d1701538b35bd08ce7/homeassistant/components/bluetooth/api.py#L118-L121), I'll keep active scanning, as I cannot test if passive scanning is enough and it has currently no benefit.

**Comment by v6ak** [comment]
Just pushed a new version:

* Doesn't fail when toothbrush isn't available during boot.
* Availability notification now seem to be reliable, thanks to async_rediscover_address.
* The library should no longer poll for toothbrushes (it doesn't reconnect without an advertisment). Benefit: it doesn't occupy connection slots.
* Attempt to work around missing battery info by reinitializing SonicareBLETB, maybe not successful.

Known issues that remain:

* Battery status isn't always updated
* In some cases, default values (e.g. battery status = 0) might propagate to entities
* Values aren't stored after HA reboot

**Comment by v6ak** [comment]
Well, I started getting issues with my BLE proxy. I have just a single proxy. It fails with the following error message:

```
2024-12-02 14:39:52.131 ERROR (MainThread) [homeassistant] Error doing job: Fatal error: protocol.data_received() call failed. (None)
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/asyncio/selector_events.py", line 1017, in _read_ready__data_received
    self._protocol.data_received(data)
  File "aioesphomeapi/_frame_helper/noise.py", line 205, in aioesphomeapi._frame_helper.noise.APINoiseFrameHelper.data_received
  File "aioesphomeapi/_frame_helper/noise.py", line 364, in aioesphomeapi._frame_helper.noise.APINoiseFrameHelper._handle_frame
  File "aioesphomeapi/connection.py", line 956, in aioesphomeapi.connection.APIConnection.process_packet
  File "/usr/local/lib/python3.12/site-packages/bleak_esphome/backend/scanner.py", line 42, in async_on_raw_advertisements
    self._async_on_advertisement(
  File "src/habluetooth/base_scanner.py", line 356, in habluetooth.base_scanner.BaseHaRemoteScanner._async_on_advertisement
  File "src/habluetooth/base_scanner.py", line 389, in habluetooth.base_scanner.BaseHaRemoteScanner._async_on_advertisement
TypeError: Expected unicode, got list
2024-12-02 14:39:52.142 WARNING (MainThread) [aioesphomeapi.connection] <name of BLE proxy> @ <IP address of BLE proxy>: Connection error occurred: Expected unicode, got list
```

Then, the device with BLE proxy keeps almost offline. It doesn't update sensor data and it cannot connect BT devices, but it can propagate advertisments and a button press.

Maybe I have an outdated software on the BLE proxy. I'll see if update fixes something.

**Comment by v6ak** [comment]
## Investigation of BLE proxies crashes

ESPhome upgrade didn't fix it.

The weirdest part of the error is that it comes from this line (not sure if the version perfectly matches, but at least the specific line is the same in HA Docker image):

https://github.com/Bluetooth-Devices/habluetooth/blob/7526057b1839b9929a802600fdb5c117465ea1ea/src/habluetooth/base_scanner.py#L389

How can this line cause a TypeError?

## Temporary workaround

I've disabled the async_rediscover_address call, which seems to cause some wrong state somewhere. Also, this call is probably not much frequently used, so it can be less tested. It seems that it has resolved the issues with BLE proxy crashes.

The drawback is that the toothbrush doesn't reconnect that reliably. Again, this is likely not a big issue in common usage scenarios (brushing twice a day), but it can be annoying when testing.

**Comment by v6ak** [comment]
Just a timeline what happens when you activate toothbrush:

1. You activate a toothbrush (move it, brush teeth, stop charging, …).
2. Toothbrush announces availability.
3. HA detects the toothbrush availability.
4. HA connects to the toothbrush and reads some data.
5. Toothbrush appears disconnected (lights are off).
6. Toothbrush appears disconnected in the addon. (It can be about a minute or two later than previous step.)
7. Toothbrush appears unavailable in Home Assistant. (It can be about 7 minutes later than the previous step! Moreover, if you activate toothbrush within this period, you probably prolong it.)

Between 6 and 7, the toothbrush is disconnected, but Home Assistant still considers it to be available. This means that when HA sees a new advertisment, it doesn't think the toothbrush becomes available. It thinks that the toothbrush is still available. As a result, it doesn't notify the addon. So, when user attempts to wake the toothbrush within this period, HA doesn't forward the advertisment to the addon. Even worse, it probably resets a timeout, so the user has actually to wait even more. (I think it is [this 5min timeout](https://github.com/Bluetooth-Devices/habluetooth/blob/main/src/habluetooth/const.py)).

Attempts to resolve the issue:

1. `bluetooth.async_set_fallback_availability_interval(hass, address, 1.0)` – doesn't do anything
2. `bluetooth.async_rediscover_address(self.hass, self.address)` when disconnected – it causes the addon to receive proper notification when the toothbrush is connected again (even between 6 and 7), but when the addon tries to reconnect between 6 and 7, it causes some weird crash of BLE proxy. I need to restart both the BLE proxy and Home Assistant in order to recover from it. Maybe it would work without BLE proxies. Weird fact: It seems that even after async_rediscover_address, the device isn't considered as unavailable.
3. Don't care about it. In typical scenarios, you probably don't need to use the toothbrush again  within ~ 7 minutes.

**Comment by Dodoooh** [comment]
Hey @v6ak 

Thank you so much for your effort—it’s brilliant! I tried setting it up today, but unfortunately, it doesn’t seem to work for me. The toothbrush does appear, but as soon as I try to add it, it just keeps loading endlessly.

How did you manage to get it working? I’ve tried everything: adding it after charging, during brushing, while shaking it, spinning in circles… nothing seems to work.

What I also noticed is that Shelly BT proxies don’t seem to function in this case (though they work fine for everything else and are set to active). To troubleshoot, I placed an ESPHome BT Proxy in the bathroom.

If there’s any log or information I can provide, please let me know—I’m happy to help in any way I can!

But really, thanks again! I’d love to get this integration up and running.

**Comment by PV-Web** [comment]
Awesome to see this topic alive again! I will soon test it with shelly BT proxy if it works on my end. Unfortunately I do not have any other bluetooth proxies available to test it with.

**Comment by v6ak** [comment]
My effort was focused on reliability of connection after the toothbrush is added, not on adding the toothbrush. However, I've tried to remove my toothbrush and add it again while having all BT adapters (except BT proxies) disabled. So, it must have connected through the BT proxy and it succeeded. (I have just one BT proxy.)

Reliability of adding a new toothbrush might have been improved as a side effect, because the addon no longer requires successful connection for initialization. But I am not sure.

If you still have issues with adding the toothbrush to HA:

* Be aware that the opportunity to connect the toothbrush is usually quite short.
* When brushing, the opportunity to connect might be longer, but I haven't tried to connect it during brushing.
* However, I usually wake the toothbrush just by charging it for few seconds. Because my toothbrush stands on a charger, which is powered through a smart socket, I can do this easily at distance.
* You might want to disable BT on your phone, as the phone might also try to connect to the toothbrush.
* Usual advices: direct visibility, short distance…


----

@Dodoooh If you can share some logs (related to Sonicare), you are welcome. Please try it with the newest version.

----

Current known issues:

* Battery status isn't always reported. This looks a bit random. However, since I've seen the same issue with the official app, I am not sure if I can do anything about it.
* Sometimes, BLE proxy kind of crashes (as described at https://github.com/GrumpyMeow/sonicare-ble-hacs/issues/14#issuecomment-2512208135 ) and I have to restart both the proxy and HA. This looks like issue in HA and/or related software (ESPHome etc.). I guess I could report an issue, but reproduction steps currently require a quite specific hardware (Sonicare toothbrush). If I had a reproduction steps for a more common hardware (e.g., Android phone or Linux-based PC), it would allow me to create a better report.
* When some value isn't loaded, you see a zero-like value, not "unavailable".
* After a toothbrush is disconnected, there is a short period (few minutes) when the toothbrush cannot connect again, as described at https://github.com/GrumpyMeow/sonicare-ble-hacs/issues/14#issuecomment-2512746859 .

**Comment by v6ak** [comment]
Observation: Battery is reported again not only HA reboot, but also after ESP32 reboot. Maybe after any reconnection between ESPHome and HA. (Would router reboot also do the trick?)

However, I am not sure if I'll continue with working on the fork:

1. It already somehow works for me, although it isn't perfect.
2. The remaining issues I am aware of don't seem to be easy to debug/solve. Still, I can somehow support someone who wants further develop my fork, if it doesn't take much time. That is, I can share my knowledge learned, but probably won't participate in deep debugging.
3. There is an alternative solution for ESP32 users: https://community.home-assistant.io/t/philips-sonicare-integration-for-bluetooth-toothbrushes/178229/27

**Comment by mtheli** [comment]
I've been working extensively on Sonicare BLE in the past months and wanted to share some findings that might help.

Looking across the open issues in this repo, a common root cause for "not able to connect" is **missing BLE bonding**. Many Sonicare models require bonding before they allow GATT access — without it, the connection simply fails or the brush disconnects immediately. This integration doesn't implement bonding, which likely affects these issues:

- #2 (HX991B) — DiamondClean 9000, requires bonding
- #3 (HX960B) — ExpertClean, requires bonding
- #8 (HX991X) — DiamondClean 9000, requires bonding
- #6 (HX9914) — requires bonding (GATT scan shows `a651fff1` transport service)
- #12 (HX962K) — ExpertClean, requires bonding

If you want to check whether your brush is bonded, you can try on the command line:

```bash
# pair the brush (press the power button first to wake it)
bluetoothctl pair <MAC>
# then check if it's bonded
bluetoothctl info <MAC> | grep Bonded
```

In the meantime, two alternative integrations have emerged:

- **[ha-sonicare-battery](https://github.com/Koky05/ha-sonicare-battery)** — event-driven battery tracking with state restoration, avoids the reconnect issues discussed here.

- **[philips_sonicare_ble](https://github.com/mtheli/philips_sonicare_ble)** — if you'd also like live brushing data, mode/intensity sensors, pressure alerts, and brush head wear tracking on top of battery. Supports BLE bonding and comes with a dedicated ESP32 BLE Bridge as an alternative to generic BT proxies. There's also a companion [Toothbrush Card](https://github.com/mtheli/toothbrush-card) for a visual dashboard. Both available via HACS.

For philips_sonicare_ble, I'm actively building out the compatibility list and would appreciate testing from anyone with a Sonicare model — especially if it didn't work with this integration. Feel free to open an issue on my repo with your model number and I'll help get it running.

@bcutter @v6ak @adheaps @PV-Web @helmerzNL @badewanne1234 @Dodoooh — in case you're still looking for a working solution.

**Comment by bcutter** [comment]
@mtheli Thanks a lot Martin, looking very promising 👍🏼

I tend to test your https://github.com/mtheli/philips_sonicare_ble integration once I find some time. Only the fact I need a dedicated ESP32 - I currently have a regular Bluetooth Proxy running closely to the Philips Sonicare, but acting as a Bluetooth repeater for other devices - is "the hair in the soup" for me because I would need to run the mentioned custom ESP32 only for the toothbrush.

I'm not that deeply into ESPs and ESPhome but I don't think it's possible to run both services (Bluetooth proxy as now plus the custom one needed for your integration) on the very same ESP32, isn't it? Maybe we shall continue that discussion in your repo anyway 😀

Thanks so far, finally more options and solutions for Philips Sonicare users 👍🏼

**Comment by Koky05** [comment]
@mtheli Thanks for your hard work, I really appreciate it! It looks like we were both working on the same task — but my Bluetooth knowledge is not as good as yours, so I only created a basic battery monitor ([ha-sonicare-battery](https://github.com/Koky05/ha-sonicare-battery)).

I've tested your integration with my two Sonicare Kids (HX6352 and HX6322) and it works perfectly — much richer data than what I had. So I've archived my repository with a redirect to your work. Great job! 👏

**Comment by mtheli** [comment]
@Koky05 Thank you so much for testing and for the kind words! Really glad it works well with both your Kids brushes — HX6352 and HX6322 are now confirmed on the compatibility list thanks to you.

Your ha-sonicare-battery was a clean solution for the battery use case — I appreciate you redirecting users to philips_sonicare_ble nonetheless. If you ever notice anything off with a future update, don't hesitate to open an issue.

**Comment by mtheli** [comment]
@bcutter Thanks for the kind words! And good question.

Unfortunately, `bluetooth_proxy` and the Sonicare component **can't run on the same ESP32** right now. There's a [bug in the Bluedroid BLE stack](https://github.com/mtheli/philips_sonicare_ble/blob/master/docs/KNOWN_ISSUES.md#bluetooth_proxy-is-not-compatible-with-the-sonicare) that causes the ESP32 to crash during GATT discovery when multiple GATT client interfaces are registered — which is exactly what `bluetooth_proxy` does. A [fix](https://github.com/espressif/esp-idf/commit/d4f3517da4a81144eaa3f091848e61ec68ab3700) has been merged to ESP-IDF master but hasn't made it into a release yet, so ESPHome doesn't have it.

However, you might not need the ESP Bridge at all — the integration also supports **Direct BLE** from your Home Assistant host. If your Sonicare is within Bluetooth range, it works without any extra hardware. What model do you have? That helps me tell you what to expect.

More details in [ESP32 Bridge docs](https://github.com/mtheli/philips_sonicare_ble/blob/master/docs/ESP32_BRIDGE.md) and [Known Issues](https://github.com/mtheli/philips_sonicare_ble/blob/master/docs/KNOWN_ISSUES.md).

**Comment by apaperclip** [comment]
This is great @mtheli ! I would imagine other ble proxy components are hitting that IDF bug, maybe the esphome team could help poke that bugfix with espressif.

**Comment by badewanne1234** [comment]
Now this is absolute awesome news, thanks heaps for your effort @mtheli!
I only have Shelly BT Proxies in place, given that the BT GATT issue has already been fixed but will likely still take some time to appear in HA, I wanted to add the custom integration from @Koky05 - is it possible that you revive your ha-battery integration, @Koky05? I cant add the repo anymore, I believe its because its archived. I really only want the Sonicare battery state and with your integration, it seems that this is possible. Of course, I'll need to test if its working with the Shelly BT Proxy.

Thanks everyone for the great effort :)

**Comment by Koky05** [comment]
@badewanne1234: Try dirrect link [ha-sonicare-battery](https://github.com/Koky05/ha-sonicare-battery). But I really suggest to try @mtheli integration. I own one BT dongle (Assus BT-500) and several Shelly in BT proxy mode and it found my toothbruses and it just work with all that sensors.

**Comment by badewanne1234** [comment]
> [@badewanne1234](https://github.com/badewanne1234): Try dirrect link [ha-sonicare-battery](https://github.com/Koky05/ha-sonicare-battery). But I really suggest to try [@mtheli](https://github.com/mtheli) integration. I own one BT dongle (Assus BT-500) and several Shelly in BT proxy mode and it found my toothbruses and it just work with all that sensors.

Of course I did use that exact link, but it seems to be "official" that you can't add archived repos. 
I dont have a BT adapter nor a ESP32 Bridge - nevertheless, I tried the new Sonicare BLE integration but you need to select either a BT adapter or ESP32 bridge, so I'm out unfortunately, I could not select any of my proxies in the setup screen. I think in that matter, that your battery integration would be still great for users with only proxies like me :)

**Comment by bcutter** [comment]
> @bcutter Thanks for the kind words! And good question.
> 
> Unfortunately, `bluetooth_proxy` and the Sonicare component **can't run on the same ESP32** right now. There's a [bug in the Bluedroid BLE stack](https://github.com/mtheli/philips_sonicare_ble/blob/master/docs/KNOWN_ISSUES.md#bluetooth_proxy-is-not-compatible-with-the-sonicare) that causes the ESP32 to crash during GATT discovery when multiple GATT client interfaces are registered — which is exactly what `bluetooth_proxy` does. A [fix](https://github.com/espressif/esp-idf/commit/d4f3517da4a81144eaa3f091848e61ec68ab3700) has been merged to ESP-IDF master but hasn't made it into a release yet, so ESPHome doesn't have it.
> 
> However, you might not need the ESP Bridge at all — the integration also supports **Direct BLE** from your Home Assistant host. If your Sonicare is within Bluetooth range, it works without any extra hardware. What model do you have? That helps me tell you what to expect.
> 
> More details in [ESP32 Bridge docs](https://github.com/mtheli/philips_sonicare_ble/blob/master/docs/ESP32_BRIDGE.md) and [Known Issues](https://github.com/mtheli/philips_sonicare_ble/blob/master/docs/KNOWN_ISSUES.md).

The HA server internal Bluetooth did not work for me because it's a bit too far away and I had to bring the toothbrush closely to the server on every HA core restart to finish the integration setup, a total mess. So the ESP satellite approach is the one working way better for me (having one Bluetooth Proxy and one Shelly closely to the toothbrush), unfortunately recent integrations could not handle this.

I tried to follow the root cause in your linked references but I can't decide which is the single information to follow - **the one issue to track**. I'm aiming at waiting for the native Bluetooth Proxy to support this (or being supported by your integration). No idea if it's a few weeks or even years? You linked one 6 years old issue, unsolved - so just wondering 😀

**Comment by mtheli** [comment]
@bcutter @apaperclip Good news — I found a fix for the `bluetooth_proxy` crash and verified it today.

The crash is caused by a NULL pointer bug in ESP-IDF's Bluedroid BLE stack ([espressif/esp-idf#4971](https://github.com/espressif/esp-idf/issues/4971)). A fix exists in ESP-IDF but hasn't been released yet. I've created a pre-build patch script that applies the fix at compile time — tested and confirmed working with `bluetooth_proxy` + the [philips_sonicare_ble](https://github.com/mtheli/philips_sonicare_ble) ESP32 Bridge component on the same ESP32.

This means you can run both `bluetooth_proxy` (for all your other BLE devices) **and** the Sonicare component on a single ESP32 — no separate device needed.

Here's what to do:

1. Copy [`bluedroid_null_fix.py`](https://github.com/mtheli/philips_sonicare_ble/blob/master/esphome/bluedroid_null_fix.py) to your ESPHome config directory (`/config/esphome/`)
2. Add this to your YAML under `esphome:`:
   ```yaml
   esphome:
     platformio_options:
       extra_scripts:
         - pre:/config/esphome/bluedroid_null_fix.py
   ```
3. Add `bluetooth_proxy` to the same YAML:
   ```yaml
   bluetooth_proxy:
     active: true
   ```
4. Do a **clean build** ("Clean Build Files") before flashing

The patch is idempotent and will show `[bluedroid-fix] Patched: bta_gattc_act.c / bta_gattc_cache.c` in the compile log.

I've also filed [esphome/esphome#15783](https://github.com/esphome/esphome/issues/15783) to get the fix cherry-picked upstream. Once that happens, the patch script won't be needed anymore.

Details and example YAMLs: [Known Issues — Workaround](https://github.com/mtheli/philips_sonicare_ble/blob/master/docs/KNOWN_ISSUES.md#workaround)

**Comment by mtheli** [comment]
Here's what it looks like in practice — Philips Sonicare (active GATT connection via the ESP Bridge component) and Oral-B (passive BLE advertisement via `bluetooth_proxy`) running on the **same ESP32** simultaneously:

<img width="923" height="476" alt="Sonicare + Oral-B on the same ESP32" src="https://github.com/user-attachments/assets/f50dd462-abda-4080-ba8b-26bf0fc43a11" />

**Comment by mtheli** [comment]
@badewanne1234 Unfortunately, Shelly BT Proxies won't work with any Sonicare integration that needs active GATT connections — Shelly devices [only support passive BLE scanning](https://www.home-assistant.io/integrations/shelly/) (advertisements), not active GATT connections.

For the [philips_sonicare_ble](https://github.com/mtheli/philips_sonicare_ble) integration, you have two options:

1. **USB Bluetooth adapter** on your HA host (like the Asus BT-500 that @Koky05 uses) — works with Direct BLE, no ESP32 needed
2. **ESP32 as BLE Bridge** (e.g. M5Stack Atom Lite, ~€10) — placed near the toothbrush, connects over WiFi to HA. With the [`bluedroid_null_fix.py`](https://github.com/mtheli/philips_sonicare_ble/blob/master/esphome/bluedroid_null_fix.py) patch, it can also serve as `bluetooth_proxy` for your other BLE devices at the same time — so it would effectively replace your Shelly proxy in that location.

**Comment by mtheli** [comment]
Quick update on the `bluetooth_proxy` fix I [mentioned above](https://github.com/GrumpyMeow/sonicare-ble-hacs/issues/14#issuecomment-4262668806).

Espressif has now backported the fix upstream:

- Merged to `release/v5.5`, `release/v5.3`, `release/v5.2` ([confirmation from Weijian-Espressif](https://github.com/esphome/esphome/issues/15783#issuecomment-4264586477))
- `release/v5.4` and `release/v6.0` are waiting on their current release freeze

**The catch:** the fix is in the release branches but not in any tagged release yet. The most recent ESP-IDF tag (v5.5.4, 2026-03-24) was cut one day before the backport landed. Based on the v5.5.x release cadence (~5–8 weeks between tags), I'd expect v5.5.5 around mid-May, followed by an ESPHome release a few weeks after that.

**Until then:** my pre-build patch script ([bluedroid_null_fix.py](https://github.com/mtheli/philips_sonicare_ble/blob/master/esphome/bluedroid_null_fix.py)) remains the workaround — it applies the same NULL-check changes at compile time and is idempotent (won't patch twice).

**Once ESPHome ships with ESP-IDF v5.5.5+**, the `extra_scripts` entry can be removed from the YAML. Parts of the script will naturally become no-ops against the patched source; the rest would only add redundant NULL checks, but it's cleaner to drop the workaround once it's obsolete.

I'll post another update here when v5.5.5 is tagged.

**Comment by bcutter** [comment]
> Do a **clean build** ("Clean Build Files") before flashing

Is this plus performing an update (2026.3.3 to 2026.4.0 is offered) sufficient?
I don't see the `[bluedroid-fix] Patched: bta_gattc_act.c / bta_gattc_cache.c` line in the logs of the ESP.

**Comment by mtheli** [comment]
@bcutter 
You are right. Please choose "Clean All Files"
<img width="258" height="468" alt="Image" src="https://github.com/user-attachments/assets/9e90c5b5-dfa3-4acb-9408-df843d4db9d6" />

**Comment by bcutter** [comment]
> @bcutter 
> You are right. Please choose "Clean All Files"
> <img width="258" height="468" alt="Image" src="https://github.com/user-attachments/assets/9e90c5b5-dfa3-4acb-9408-df843d4db9d6" />

Done, did a new install - still no line in the logs.
Am I the first one actually trying this workaround? Am I holding it wrong? :-)

**Comment by mtheli** [comment]
> > [@bcutter](https://github.com/bcutter)
> > You are right. Please choose "Clean All Files"
> > <img alt="Image" width="258" height="468" src="https://private-user-images.githubusercontent.com/1991952/580335516-9e90c5b5-dfa3-4acb-9408-df843d4db9d6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzY1ODMxMjcsIm5iZiI6MTc3NjU4MjgyNywicGF0aCI6Ii8xOTkxOTUyLzU4MDMzNTUxNi05ZTkwYzViNS1kZmEzLTRhY2ItOTQwOC1kZjg0M2Q0ZGI5ZDYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDQxOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA0MTlUMDcxMzQ3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmM1MjA2Y2U2ZWQ3NDhiZTViNjY1Y2YwM2FiMDhlMTlhMjAyZmZkMDNmZDI5YjM2OWUzYTViZjdmNzMxYzEwNSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmcmVzcG9uc2UtY29udGVudC10eXBlPWltYWdlJTJGcG5nIn0.FD8BVzK-m5X8_YwaedjOHfS9zH43JRCpmn-Az5IaykQ">
> 
> Done, did a new install - still no line in the logs. Am I the first one actually trying this workaround? Am I holding it wrong? :-)

No. I have also tried the workaround and it worked. Where have you tried to find the log output? It is a compile time output and not a runtime log output from the ESP.

**Comment by bcutter** [comment]
Oh that was the important part. I expected it to show up in the live  (runtime) log. So it's kind of difficult to ultimately check if the fix is in place if you don't backup and parse the compile log. For now I just assume it is in place and continue to set up the integration.

**Comment by npike** [comment]
@mtheli works great with sonicare kids brush, thank you!

**Comment by mtheli** [comment]
**Update: the workaround above is no longer needed.** The underlying Bluedroid NULL-pointer fix shipped in ESP-IDF 5.5.5, and **ESPHome 2026.7.1** bundles it. I've verified this on the same hardware that used to crash within seconds — `bluetooth_proxy` plus GATT service discovery now runs fine on a completely unpatched build.

So: update the ESPHome builder to 2026.7.1 or newer, rebuild and reflash the proxy, and the `bluedroid_null_fix.py` / `extra_scripts:` wiring can be dropped from the YAML. If you're stuck on an older ESPHome for some reason, the patch from my comment above still works — it's a no-op on fixed builds.
